// @sap/cds/lib/core/entities.js
const { struct, type } = require ('./classes')

class entity extends struct   {
  is (kind) { return kind === 'entity' || kind === 'view' && !!this.query || super.is(kind) }
  get keys() {
    return this.own('_keys') || this.set('_keys', this._elements (e => e.key))
  }
  get associations() {
    return this.own('_associations') || this.set('_associations', this._elements (e => e instanceof Association))
  }
  get compositions() {
    return this.own('_compositions') || this.set('_compositions', this._elements (e => e instanceof Composition))
  }
  get drafts() {
    return this.own('_drafts') || this.set('_drafts',
      this.elements.HasDraftEntity && { name: this.name + '_drafts', keys: this.keys }
    )
  }
  _elements (filter) {
    const {elements} = this, dict={}; let any
    for (let e in elements) if (filter(elements[e])) (any = dict)[e] = elements[e]
    return any
  }
}

class Association extends type {

  is(kind) { return kind === 'Association' || super.is(kind) }
  get isAssociation(){ return true }

  get is2many() { return !this.is2one }
  get is2one() {
    let c = this.cardinality
    return !c || c.max === 1 || (!c.max && !c.targetMax) || c.targetMax === 1
  }

  set foreignKeys(k) { this.set('foreignKeys', k) }
  get foreignKeys() {
    const keys = this.keys; if (!keys) return this.foreignKeys = undefined
    const foreignKeys = {}
    for (const k of keys) {
      const el = k.ref.reduce((target,n)=> target.elements[n], this._target)
      const as = k.as || k.ref[k.ref.length-1]
      foreignKeys [as] = el
    }
    return this.foreignKeys = foreignKeys
  }

  set keys(k) { super.keys = k }
  get keys() {
    if (this.on || this.is2many || !this._target) return this._keys = undefined
    const keys=[], tks = this._target.keys
    for (let k in tks) keys.push({ ref: [tks[k].name] })
    return this.keys = keys
  }
}


class Composition extends Association {

  is(kind) { return kind === 'Composition' || super.is(kind) }
  get isComposition(){ return true }
  get isManagedComposition(){ return this._target && this._target.kind !== 'entity' }

}


module.exports = { entity, Association, Composition }
