// @sap-cloud-sdk/core/dist/odata-v2/uri-conversion/odata-uri.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oDataUriV2 = exports.oDataUri = void 0;
var odata_common_1 = require("../../odata-common");
var get_expand_1 = require("./get-expand");
var get_select_1 = require("./get-select");
var uri_value_converter_1 = require("./uri-value-converter");
var getFilter = (0, odata_common_1.createGetFilter)(uri_value_converter_1.uriConverter).getFilter;
var getResourcePathForKeys = (0, odata_common_1.createGetResourcePathForKeys)(uri_value_converter_1.uriConverter).getResourcePathForKeys;
var convertToUriFormat = uri_value_converter_1.uriConverter.convertToUriFormat;
/**
 * Instance of the [[ODataUri]] conversion interface for OData v2.
 */
exports.oDataUri = {
    getExpand: get_expand_1.getExpand,
    getFilter: getFilter,
    getEntityKeys: odata_common_1.getEntityKeys,
    getOrderBy: odata_common_1.getOrderBy,
    getResourcePathForKeys: getResourcePathForKeys,
    getSelect: get_select_1.getSelect,
    convertToUriFormat: convertToUriFormat
};
exports.oDataUriV2 = exports.oDataUri;
