// @sap/cds-dk/lib/init/template/samples/index.js
const path = require('path');

const TemplateBase = require('../templateBase');

const commandUtil = require('../../util/commandUtil');
const mvnArchetypes = require('../../util/mvnArchetypeUtil');
const term = require('../../../util/term');

const { PROJECT_TYPE, OPTION_JAVA, URLS } = require('../../constants');



module.exports = class SamplesTemplate extends TemplateBase {
    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname);
    }

    async checkEnabled() {
        return true;
    }

    async run() {
        switch (await this.getProjectType()) {
            case PROJECT_TYPE.java:
                return await this._addJavaSamples();

            case PROJECT_TYPE.nodejs:
                return await this._addNodejsSamples();

            case PROJECT_TYPE.unknown:
            default:
            // issue warning?
        }
    }

    async _addJavaSamples() {
        // only execute if not add via cds init --add java, otherwise samples would be added twice
        // do not check in checkEnabled since log output "> adding samples" wouldn't appear
        if (!this.options.add.has(OPTION_JAVA)) {
            // keep sequence since plugin needs samples to generate test classes
            const mvnAddSamplesArgs = mvnArchetypes.getAddSamplesCmdArgs(this.options);
            try {
                await commandUtil.spawnCommand('mvn', mvnAddSamplesArgs, {
                    cwd: this.projectPath
                }, this.logger);

                const mvnAddTestsArgs = mvnArchetypes.getAddTestsCmdArgs(this.options);
                await commandUtil.spawnCommand('mvn', mvnAddTestsArgs, {
                    cwd: this.projectPath
                }, this.logger);
            } catch (err) {
                if (err.code === 'ENOENT' && err.path === 'mvn') {
                    throw new Error(`Maven executable 'mvn' not found, follow ${term.info(URLS.MAVEN_INSTALL_HELP)} and install Maven on your machine.`);
                }
                throw err;
            }
        }
    }

    async _addNodejsSamples() {
        const folders = (await this.getEnv()).folders;
        const dbFolder = path.join(this.projectPath, folders.db);
        await this.templateUtil.copyFiles('nodejs/db', dbFolder, {}, this.options.force);

        const srvFolder = path.join(this.projectPath, folders.srv);
        await this.templateUtil.copyFiles('nodejs/srv', srvFolder, {
            dbFolder: folders.db.replace(/[\\/]+$/, '')
        }, this.options.force);
    }

    async finalize() {
    }
}
