// @sap-cloud-sdk/core/dist/odata-v2/request-builder/response-transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformReturnValueForEdmTypeListV2 = exports.transformReturnValueForEdmTypeList = exports.transformReturnValueForEdmTypeV2 = exports.transformReturnValueForEdmType = exports.transformReturnValueForComplexTypeListV2 = exports.transformReturnValueForComplexTypeList = exports.transformReturnValueForComplexTypeV2 = exports.transformReturnValueForComplexType = exports.transformReturnValueForEntityListV2 = exports.transformReturnValueForEntityList = exports.transformReturnValueForEntityV2 = exports.transformReturnValueForEntity = exports.transformReturnValueForUndefinedV2 = exports.transformReturnValueForUndefined = void 0;
var entity_deserializer_1 = require("../entity-deserializer");
var response_data_accessor_1 = require("./response-data-accessor");
/* eslint-disable valid-jsdoc */
/**
 * @hidden
 */
function transformReturnValueForUndefined(data, builderFn) {
    return builderFn(data);
}
exports.transformReturnValueForUndefined = transformReturnValueForUndefined;
exports.transformReturnValueForUndefinedV2 = transformReturnValueForUndefined;
/**
 * @hidden
 */
function transformReturnValueForEntity(data, entityConstructor) {
    return (0, entity_deserializer_1.deserializeEntity)((0, response_data_accessor_1.getSingleResult)(data), entityConstructor).setOrInitializeRemoteState();
}
exports.transformReturnValueForEntity = transformReturnValueForEntity;
exports.transformReturnValueForEntityV2 = transformReturnValueForEntity;
/**
 * @hidden
 */
function transformReturnValueForEntityList(data, entityConstructor) {
    return (0, response_data_accessor_1.getCollectionResult)(data).map(function (entityJson) {
        return (0, entity_deserializer_1.deserializeEntity)(entityJson, entityConstructor).setOrInitializeRemoteState();
    });
}
exports.transformReturnValueForEntityList = transformReturnValueForEntityList;
exports.transformReturnValueForEntityListV2 = transformReturnValueForEntityList;
/**
 * @hidden
 */
function transformReturnValueForComplexType(data, builderFn) {
    return builderFn((0, response_data_accessor_1.getSingleResult)(data));
}
exports.transformReturnValueForComplexType = transformReturnValueForComplexType;
exports.transformReturnValueForComplexTypeV2 = transformReturnValueForComplexType;
/**
 * @hidden
 */
function transformReturnValueForComplexTypeList(data, builderFn) {
    return (0, response_data_accessor_1.getCollectionResult)(data).map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeList = transformReturnValueForComplexTypeList;
exports.transformReturnValueForComplexTypeListV2 = transformReturnValueForComplexTypeList;
/**
 * @hidden
 */
function transformReturnValueForEdmType(data, builderFn) {
    return builderFn((0, response_data_accessor_1.getSingleResult)(data));
}
exports.transformReturnValueForEdmType = transformReturnValueForEdmType;
exports.transformReturnValueForEdmTypeV2 = transformReturnValueForEdmType;
/**
 * @hidden
 */
function transformReturnValueForEdmTypeList(data, builderFn) {
    return (0, response_data_accessor_1.getCollectionResult)(data).map(builderFn);
}
exports.transformReturnValueForEdmTypeList = transformReturnValueForEdmTypeList;
exports.transformReturnValueForEdmTypeListV2 = transformReturnValueForEdmTypeList;
