// @sap-cloud-sdk/core/dist/connectivity/scp-cf/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.customAttributes = exports.userFromJwt = exports.mappingUserFields = exports.isIdenticalTenant = exports.tenantFromJwt = exports.mappingTenantFields = void 0;
__exportStar(require("./jwt"), exports);
__exportStar(require("./cache"), exports);
__exportStar(require("./legacy/client-credentials-token"), exports);
__exportStar(require("./legacy/xsuaa-service"), exports);
__exportStar(require("./client-credentials-token-cache"), exports);
__exportStar(require("./connectivity-service"), exports);
__exportStar(require("./connectivity-service-types"), exports);
__exportStar(require("./destination"), exports);
__exportStar(require("./environment-accessor"), exports);
__exportStar(require("./environment-accessor-types"), exports);
__exportStar(require("./resilience-options"), exports);
__exportStar(require("./token-accessor"), exports);
__exportStar(require("./verification-keys"), exports);
__exportStar(require("./xsuaa-service-types"), exports);
__exportStar(require("./protocol"), exports);
__exportStar(require("./authorization-header"), exports);
__exportStar(require("./get-protocol"), exports);
__exportStar(require("./header-builder-for-destination"), exports);
var tenant_1 = require("./tenant");
Object.defineProperty(exports, "mappingTenantFields", { enumerable: true, get: function () { return tenant_1.mappingTenantFields; } });
Object.defineProperty(exports, "tenantFromJwt", { enumerable: true, get: function () { return tenant_1.tenantFromJwt; } });
Object.defineProperty(exports, "isIdenticalTenant", { enumerable: true, get: function () { return tenant_1.isIdenticalTenant; } });
var user_1 = require("./user");
Object.defineProperty(exports, "mappingUserFields", { enumerable: true, get: function () { return user_1.mappingUserFields; } });
Object.defineProperty(exports, "userFromJwt", { enumerable: true, get: function () { return user_1.userFromJwt; } });
Object.defineProperty(exports, "customAttributes", { enumerable: true, get: function () { return user_1.customAttributes; } });
