// @sap/eslint-plugin-cds/lib/rules/latest-cds-version.js
const cp = require("child_process");
const semver = require("semver");

module.exports = {
  meta: {
    docs: {
      description: "Checks whether the latest `@sap/cds` version is being used.",
      category: "Environment",
      version: "1.0.4",
    },
    type: "suggestion",
    hasSuggestions: true,
    messages: {
      latestCDSVersion: `A newer CDS version is available!`,
    },
  },
  create: function (context) {
    return { all: check_latest_cds_version };

    function check_latest_cds_version() {
      let cdsVersions;
      let e = context.cds.environment;
      if (!e) {
        e = cp
          .execSync(`npm outdated @sap/cds --json`, {
            cwd: process.cwd(),
            stdio: "pipe",
          })
          .toString();
      }
      if (e && e["@sap/cds"]) {
        cdsVersions = e["@sap/cds"];
        // If current cds version is not the latest
        if (Object.keys(cdsVersions).length !== 0 && !semver.satisfies(cdsVersions.latest, cdsVersions.current)) {
          return {
            messageId: "latestCDSVersion",
          };
        }
      }
    }
  },
};
