// @sap/eslint-plugin-cds/lib/rules/no-join-on-draft-enabled-entities.js
module.exports = {
  meta: {
    docs: {
      description: `Draft-enabled entities shall not be used in views that make use of \`JOIN\`.`,
      category: "Model Validation",
      recommended: true,
      version: "2.2.1",
    },
    severity: "warn",
    type: "suggestion",
    messages: {
      noJoinOnDraftEnabledEntities: `Do not use draft-enabled entities in views that make use of \`JOIN\`.`,
    },
  },
  create: function (context) {

    return { entity: check_nojoin_draftenabled }

    function check_nojoin_draftenabled(e) {
      if (e["@odata.draft.enabled"]) {
        if (e.query.SELECT.from.join) {
          const location = e.query.$location;
          if (context.sourcecode.lines[location.line - 1]) {
            const endCol = context.sourcecode.lines[location.line - 1].length;
            const loc = {
              start: { line: location.line, column: location.col - 1 },
              end: { line: location.line, column: endCol },
            };
            return {
              messageId: "noJoinOnDraftEnabledEntities",
              loc,
              file: e.$location.file,
            };
          }
        }
      }
    }

  },
};
