// @sap/eslint-plugin-cds/lib/rules/sql-cast-suggestion.js
/* eslint-disable no-undef */
module.exports = {
  meta: {
    docs: {
      description: "Should make suggestions for possible missing SQL casts.",
      category: "Model Validation",
      recommended: true,
      version: "1.0.8",
    },
    severity: "warn",
    type: "suggestion",
    hasSuggestions: true,
    messages: {
      missingSQLCast: "Potential issue - Missing SQL cast for column expression?",
    },
  },
  create: function (context) {
    return { view: check_sql_cast };

    function check_sql_cast(v) {
      const reports = [];
      if (v.query && v.query.SET) {
        for (const { SELECT } of v.query.SET.args) {
          // Only in UNION cases?
          for (const each of SELECT.columns || []) {
            const { xpr, cast, $location: location } = each;
            if (cast && xpr) {
              if (xpr[0].xpr && xpr[0].cast) {
                continue;
              } else {
                if (context.sourcecode.lines[location.line - 1]) {
                  const endCol = context.sourcecode.lines[location.line - 1].length;
                  const loc = {
                    start: { line: location.line, column: location.col - 1 },
                    end: { line: location.line, column: endCol },
                  };
                  reports.push({
                    messageId: "missingSQLCast",
                    loc,
                    file: location.file,
                  });
                }
              }
            }
          }
        }
      }
      if (reports.length > 0 ) return reports;    
    }
    
  },
};
