// @sap-cloud-sdk/core/dist/odata-v2/entity-serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeEntityNonCustomFieldsV2 = exports.serializeComplexTypeV2 = exports.serializeEntityV2 = exports.entitySerializerV2 = exports.serializeEntityNonCustomFields = exports.serializeComplexType = exports.serializeEntity = exports.entitySerializer = void 0;
var entity_serializer_1 = require("../odata-common/entity-serializer");
var payload_value_converter_1 = require("./payload-value-converter");
/**
 * Entity serializer instance for v2 entities.
 * See [[EntitySerializer]] for the provided methods.
 */
exports.entitySerializer = (0, entity_serializer_1.entitySerializer)(payload_value_converter_1.tsToEdm);
exports.entitySerializerV2 = exports.entitySerializer;
exports.serializeEntity = exports.entitySerializer.serializeEntity;
exports.serializeEntityV2 = exports.serializeEntity;
exports.serializeComplexTypeV2 = exports.serializeEntity;
exports.serializeComplexType = exports.entitySerializer.serializeComplexType;
exports.serializeEntityNonCustomFields = exports.entitySerializer.serializeEntityNonCustomFields;
exports.serializeEntityNonCustomFieldsV2 = exports.serializeEntityNonCustomFields;
