// @sap-cloud-sdk/core/dist/odata-v2/entity-deserializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entityDeserializerV2 = exports.deserializeComplexTypeV2 = exports.deserializeEntityV2 = exports.deserializeComplexType = exports.deserializeEntity = exports.entityDeserializer = void 0;
var entity_deserializer_1 = require("../odata-common/entity-deserializer");
var payload_value_converter_1 = require("./payload-value-converter");
var extract_odata_etag_1 = require("./extract-odata-etag");
var response_data_accessor_1 = require("./request-builder/response-data-accessor");
/**
 * Entity deserializer instance for v2 entities.
 * See [[EntityDeserializer]] for the provided methods.
 */
exports.entityDeserializer = (0, entity_deserializer_1.entityDeserializer)(payload_value_converter_1.edmToTs, extract_odata_etag_1.extractODataEtag, response_data_accessor_1.getLinkedCollectionResult);
exports.entityDeserializerV2 = exports.entityDeserializer;
exports.deserializeEntity = exports.entityDeserializer.deserializeEntity;
exports.deserializeEntityV2 = exports.deserializeEntity;
exports.deserializeComplexType = exports.entityDeserializer.deserializeComplexType;
exports.deserializeComplexTypeV2 = exports.deserializeComplexType;
