// @sap/cds-dk/lib/init/template/auditlog/index.js
const fs = require('fs');
const path = require('path');
const TemplateBase = require('../templateBase');
const ProjectReader = require('../../util/projectReader');
const { srvNode, srvJava, auditlog } = require('../_merging/existences');

module.exports = class AuditlogTemplate extends TemplateBase {
    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname);
        this.projectReader = new ProjectReader(this);
    }

    async runMTAMerging() {
        const projectDescriptor = await this.projectReader.read(this.options);
        const mtaYAMLPath = path.join(this.projectPath, 'mta.yaml');
        if (!fs.existsSync(mtaYAMLPath)) return;
        const srv = projectDescriptor.cap.isNodejs ? srvNode : srvJava
        await this.templateUtil.mergeYAML(
            mtaYAMLPath,
            `${__dirname}/files/mta.yaml.hbs`,
            projectDescriptor,
            {
                existences: [srv, auditlog],
                relationships: [{
                    in: srv,
                    inKeyPath: ["requires"],
                    into: "name",
                    existences: [auditlog],
                    existenceKeyPath: ["name"],
                }],
            }
        );
    }

    async run() {
        await this.runMTAMerging()
    }
}
