// @sap/cds-dk/lib/init/template/lint/files/.eslint/rules/no-entity-moo.js
module.exports = require("@sap/eslint-plugin-cds/lib/api").createRule({
  meta: {
    docs: {
      description: `Demonstrates how to implement a model validation check.

           __________________-_______
          < Entity 'Moo' not allowed! >
           --------------------------
                  \   ^__^
                   \  (oo)\_______
                      (__)\       )\/\
                          ||----w |
                          ||     ||`,
      version: "1.0.0",
    },
    fixable: "code",
  },
  create: function (context) {
    const m = context.cds.model;
    m.forall("entity", (d) => {
      const entityName = d.name.split(".").pop();
      if (entityName === "Moo") {
        const loc = context.cds.getLocation(entityName, d);
        const fix = (fixer) => {
          const rangeEnd = context.sourcecode.getIndexFromLoc({
            line: loc.end.line,
            column: loc.end.column,
          });
          return fixer.insertTextAfterRange([0, rangeEnd], "n");
        };
        context.report({
          message: "Entity 'Moo' not allowed!",
          loc,
          fix,
          file: d.$location.file,
        });
      }
    });
  },
})
