// @sap/eslint-plugin-cds/lib/index.js
/**
 * Custom ESLint plugin:
 * https://eslint.org/docs/developer-guide/working-with-plugins
 * This file exposes our plugins ESLint configuration, which must:
 * - Expose any 'configs' for prescribed rule configuration bundles
 *     (i.e. "recommended"). See shareable configs:
 *     https://eslint.org/docs/developer-guide/shareable-configs
 * - Expose any 'globals' for use in ESLint
 * - Expose any 'processors' for use in ESLint
 * - Expose any 'rules' for use in ESLint
 */

const path = require("path");
const processor = require("./processor");

const { FILES, GLOBALS, PLUGIN_NAME, PROCESSOR_NAME } = require("./constants");
const { getRules } = require("./utils/rules");

const rules = getRules(path.join(__dirname, "rules"));

function _getConfig(configName) {
    return {
        globals: GLOBALS,
        plugins: [PLUGIN_NAME],
        overrides: [
            {
                files: FILES,
                processor: PROCESSOR_NAME,
            },
        ],
        rules: rules[configName],
    };
}

module.exports = {
    configs: {
        recommended: _getConfig("recommended"),
        all: _getConfig("all"),
    },
    processors: {
        cds: processor,
    },
    rules: rules.sources,
};
