// @sap/cds-dk/lib/import/message.js
"use strict";

const messages = {
	INVALID_EDMX_METADATA: 'The OData model is not valid. Specify the correct OData model.',
	INVALID_OUTPUT_FILE: 'Please provide an output file with one of these [cds, csn, json] file extension.',
	MISSING_ENTITIES: 'There are no entities in the OData model. Specify the correct OData model in the provided file.',
	SPECIFY_INPUT_FILE: 'Specify an input file that exists.',
	SPECIFY_OUTPUT_FILE: 'Specify an output file or correct the output file path provided.',
	MISSING_ENTITY_CONTAINER: 'There is no entity container in the OData model. Specify the correct OData model.',
	MISSING_ENTITY_SETS: 'There are no entity sets in the OData model. Specify the correct OData model.',
	INVALID_INPUT_FILE: 'Input file provided is not valid. Please provide a file with either .edmx or .xml extension.',
	FILE_MODIFIED: 'The file with the same name exists and is been modified. Use --force flag to forcefully overwrite the file.',
	OUTPUT_FILE_MISMATCH: 'Specify the file extension that --as option supports.',
	INVALID_AS_OPTION: 'For --as option, please specify one of these output file format [cds, csn, json].',
	INCORRECT_EXTENSION: 'If --as option is not provided, then please provide either .json or .csn output file extension.',
	INCORRECT_INPUT_NAMESPACES: 'Invalid value for --include-namespaces option. Specify namespaces as comma-separated list or "*"',
	MISSING_ASSOCIATION_SETS: 'Input file provided has missing Association Sets. Please check the edmx file.',
	COLLECTION_IN_KEY: 'Key Property of an Entity cannot be of the type Collection.'
};

function getMessages() {
	return messages;
}

module.exports = {
	getMessages
};
